File: Problems/117/main.go
package main

import "container/list"

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

func main() {}

func connect(root *Node) *Node {
	// BFS ..., using node value of -101 as ending signal
	queue := list.New()
	queue.PushBack(root)

	for queue.Len() != 0 {
		n := queue.Len()

		cur := queue.Front()
		var node *Node
		for n > 0 {
			// IMPORTANT !!
			// If the type assertion holds, the value of the expression is the value stored in x and its type is T.
			// https://go.dev/ref/spec#Type_assertions
			// if cur.Value != nil {
			if cur.Value.(*Node) != nil {
				// Update Next Pointer
				if node != nil {
					node.Next = cur.Value.(*Node)
				}
				node = cur.Value.(*Node)

				queue.PushBack(node.Left)
				queue.PushBack(node.Right)
			}
			toRemove := cur
			cur = cur.Next()
			queue.Remove(toRemove)
			n--
		}
	}

	return root
}

File: Problems/117/main.cpp
#include <vector>
#include <iostream>
#include <string>
#include <map>
#include <stack>
#include <algorithm>
#include <numeric>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <functional>
#include <sstream>
#include <set>

using namespace std;

class Node
{
public:
    int val;
    Node *left;
    Node *right;
    Node *next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}
    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}
    Node(int _val, Node *_left, Node *_right, Node *_next)
        : val(_val), left(_left), right(_right), next(_next) {}
};

class Solution
{
public:
    Node *connect(Node *root)
    {
        if (!root)
            return root;
        queue<Node *> q;

        q.push(root);
        while (!q.empty())
        {
            int len = q.size();
            while (len--)
            {
                auto n = q.front();
                q.pop();
                n->next = len == 0 ? nullptr : q.front();
                if (n->left)
                    q.push(n->left);
                if (n->right)
                    q.push(n->right);
            }
        }

        return root;
    }
};
